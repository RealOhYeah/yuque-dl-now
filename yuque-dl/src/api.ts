import { env } from 'node:process'
import axios from 'axios'
import { randUserAgent } from './utils'
import { DEFAULT_COOKIE_KEY, DEFAULT_DOMAIN } from './constant'

import type {
  ArticleResponse,
  KnowledgeBase,
  GetHeaderParams,
  IReqHeader,
  TGetKnowledgeBaseInfo,
  TGetMdData
} from './types'
import type { AxiosRequestConfig } from 'axios'

function getHeaders(params: GetHeaderParams): IReqHeader {
  const { key = DEFAULT_COOKIE_KEY, token } = params
  const headers: IReqHeader = {
    'user-agent': randUserAgent({
      browser: 'chrome',
      device: 'desktop'
    })
  }
  if (token) headers.cookie = `${key}=${token};`
  return headers
}

export function genCommonOptions(params: GetHeaderParams): AxiosRequestConfig {
  const config: AxiosRequestConfig = {
    headers: getHeaders(params),
    beforeRedirect: (options) => {
      // 语雀免费非企业空间会重定向如: www.yuque.com -> gxr404.yuque.com
      // 此时axios自动重定向并不会带上cookie
      options.headers = {
        ...(options?.headers || {}),
        ...getHeaders(params)
      }
    }
  }
  if (env.NODE_ENV === 'test') {
    config.proxy = false
  }
  return config
}


/** 获取知识库数据信息 */
export const getKnowledgeBaseInfo: TGetKnowledgeBaseInfo = (url, headerParams) => {
  const knowledgeBaseReg = /decodeURIComponent\("(.+)"\)\);/m
  return axios.get<string>(url, genCommonOptions(headerParams))
    .then(({data = '', status}) => {
      if (status === 200) return data
      return ''
    })
    .then(html => {
      const data = knowledgeBaseReg.exec(html) ?? ''
      if (!data[1]) return {}
      const jsonData: KnowledgeBase.Response = JSON.parse(decodeURIComponent(data[1]))
      if (!jsonData.book) return {}
      const info = {
        bookId: jsonData.book.id,
        bookSlug: jsonData.book.slug,
        tocList: jsonData.book.toc || [],
        bookName: jsonData.book.name || '',
        bookDesc: jsonData.book.description || '',
        host: jsonData.space?.host || DEFAULT_DOMAIN,
        imageServiceDomains: jsonData.imageServiceDomains || []
      }
      return info
    }).catch((e) => {
      // console.log(e.message)
      const errMsg = e?.message ?? ''
      if (!errMsg) throw new Error('unknown error')
      const netErrInfoList = [
        'getaddrinfo ENOTFOUND',
        'read ECONNRESET',
        'Client network socket disconnected before secure TLS connection was established'
      ]
      const isNetError = netErrInfoList.some(netErrMsg => errMsg.startsWith(netErrMsg))
      if (isNetError) {
        throw new Error('请检查网络(是否正常联网/是否开启了代理软件)')
      }
      throw new Error(errMsg)
    })
}


export const getDocsMdData: TGetMdData = (params, isMd = true) => {
  const { articleUrl, bookId, token, key, host = DEFAULT_DOMAIN } = params
  let apiUrl = `${host}/api/docs/${articleUrl}`
  const queryParams: any = {
    'book_id': String(bookId),
    'merge_dynamic_data': String(false)
    // plain=false
    // linebreak=true
    // anchor=true
  }
  if (isMd) queryParams.mode = 'markdown'
  const query = new URLSearchParams(queryParams).toString()
  apiUrl = `${apiUrl}?${query}`
  return axios.get<ArticleResponse.RootObject>(apiUrl, genCommonOptions({token, key}))
    .then(({data, status}) => {
      const res = {
        apiUrl,
        httpStatus: status,
        response: data
      }
      return res
    })
}
