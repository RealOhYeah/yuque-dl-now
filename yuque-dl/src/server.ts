import { createServer } from 'vitepress'
import { mkdir, writeFile, readFile, readdir, stat, access } from 'node:fs/promises'
import { join, resolve, sep } from 'node:path'

let restartPromise: Promise<void> | undefined

export async function runServer(root: string) {
  const rootPath = resolve(root)
  if (!await fileExists(rootPath)) {
    throw new Error('server root not found')
  }
  await createVitePressConfig(rootPath)
  const createDevServer = async () => {
    const server = await createServer(root, {}, async () => {
      if (!restartPromise) {
        restartPromise = (async () => {
          await server.close()
          await createDevServer()
        })().finally(() => {
          restartPromise = undefined
        })
      }

      return restartPromise
    })
    await server.listen()
    server.printUrls()
    // bindShortcuts(server, createDevServer)
  }

  await createDevServer().catch((err) => {
    console.error(err)
    process.exit(1)
  })
}

async function createVitePressConfig(root: string) {
  const bookName = root.split(sep).filter(Boolean).at(-1) || 'yuque-dl'
  const vitepressPath = join(root, '/.vitepress')
  await mkdir(vitepressPath, {recursive: true})
  const vitePressConfig = join(vitepressPath, 'config.ts')
  const sidebar = await createSidebarMulti(root)
  const config = `
  export default {
    title: "${bookName}",
    themeConfig: {
      search: {
        provider: 'local'
      },
      sidebar: ${JSON.stringify(sidebar)}
    },
    vite: {
      optimizeDeps: {
        include: []
      }
    },
    markdown: {
      html: false
    }
  }
  `
  await writeFile(vitePressConfig, config)
}

async function fileExists(filename: string) {
  try {
    await access(filename)
    return true
  } catch (err) {
    if (err.code === 'ENOENT') {
      return false
    } else {
      throw err
    }
  }
}

async function createSidebarMulti (path: string) {
  const data = [] as any
  const ignoreList = ['.vitepress', 'img', 'index.md', 'progress.json', 'attachments']
  let dirList = await readdir(path)
  dirList = dirList.filter(item => !ignoreList.includes(item))
  for (const n of dirList) {
    const dirPath = join(path, n)
    const statRes = await stat(dirPath)
    if (statRes.isDirectory()) {
      const isHasIndex = await fileExists(join(dirPath,'index.md'))
      const item: any = {
        text: n,
        collapsed: true,
        items: await createSideBarItems(path, n)
      }
      if (isHasIndex) item.link = `/${n}/`
      data.push(item)
    } else {
      data.push({
        text: n.slice(0, n.lastIndexOf('.')),
        link: `/${n}`
      })
    }
  }
  return data
}

// 尝试从一个md文件中读取标题，读取到第一个 ‘# 标题内容’ 的时候返回这一行
export async function getTitleFromFile (realFileName: string): Promise<string | undefined> {
  const isExist = await fileExists(realFileName)
  if (!isExist) {
    return undefined
  }
  const fileExtension = realFileName.substring(
    realFileName.lastIndexOf('.') + 1
  )
  if (fileExtension !== 'md' && fileExtension !== 'MD') {
    return undefined
  }
  // read contents of the file
  const data = await readFile(realFileName, { encoding: 'utf-8' })
  // split the contents by new line
  const lines = data.split(/\r?\n/)
  // return title
  for (const line of lines) {
    if (line.startsWith('# ')) {
      return line.substring(2)
    }
  }
  return undefined
}

async function createSideBarItems (
  targetPath: string,
  ...reset: string[]
) {
  const collapsed = false
  const node = await readdir(join(targetPath, ...reset))

  const result = []
  for (const fname of node) {
    const curPath = join(targetPath, ...reset, fname)
    const statRes = await stat(curPath)
    if (statRes.isDirectory()) {
      const items = await createSideBarItems(join(targetPath), ...reset, fname)
      if (items.length > 0) {
        const sidebarItem: any = {
          text: fname,
          items
        }
        // vitePress sidebar option collapsed
        sidebarItem.collapsed = collapsed
        const isHasIndex = await fileExists(join(curPath, 'index.md'))
        if (isHasIndex) {
          sidebarItem.link = '/' + [...reset, fname].map(decodeURIComponent).join('/') + '/'
        }
        result.push(sidebarItem)
      }
    } else {
      // is filed
      if (
        fname === 'index.md' ||
        /^-.*\.(md|MD)$/.test(fname) ||
        !fname.endsWith('.md')
      ) {
        continue
      }
      const fileName = fname.replace(/\.md$/, '')
      const item = {
        text: fileName,
        link: '/' + [...reset.map(decodeURIComponent), `${encodeURI(fileName)}.html`].join('/')
      }
      result.push(item)
    }
  }
  return result
}
