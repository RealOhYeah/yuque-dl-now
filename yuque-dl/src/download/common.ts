import * as stream from 'node:stream'
import { promisify } from 'node:util'
import { createWriteStream } from 'node:fs'
import axios from 'axios'

import { genCommonOptions } from '../api'

interface IDownloadFileParams {
  fileUrl: string,
  savePath: string,
  token?: string
  key?: string,
  fileName: string
}

const finished = promisify(stream.finished)
export async function downloadFile(params: IDownloadFileParams) {
  const {fileUrl, savePath, token, key, fileName} = params
  return axios.get(fileUrl, {
    ...genCommonOptions({token, key}),
    responseType: 'stream'
  }).then(async response => {
    if (response.request?.path?.startsWith('/login')) {
      throw new Error(`"${fileName}" need token`)
    } else if (response.status === 200) {
      const writer = createWriteStream(savePath)
      response.data?.pipe(writer)
      return finished(writer)
        .then(() => ({
          fileUrl,
          savePath
        }))
    }
    throw new Error(`response status ${response.status}`)
  })
}
