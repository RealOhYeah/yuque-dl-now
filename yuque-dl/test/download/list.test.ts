import path from 'node:path'
import { readdirSync, readFileSync } from 'node:fs'
import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, it } from 'vitest'
import { TestTools } from '../helpers/TestTools'
import { server } from '../mocks/server'
import { downloadArticleList } from '../../src/download/list'
import { ProgressBar } from '../../src/utils'


let testTools: TestTools

describe('downloadArticle', () => {
  beforeAll(() => {
    server.listen({ onUnhandledRequest: 'error' })
  })
  afterAll(() => server.close())

  beforeEach(() => {
    testTools = new TestTools()
  })

  afterEach(() => {
    testTools.cleanup()
    server.resetHandlers()
  })
  it('should work', async () => {
    const tocList = [
      {
        type: 'TITLE',
        title: 'Title1',
        uuid: '001',
        child_uuid: '002',
        parent_uuid: ''
      },
      {
        type: 'DOC',
        title: '文档1',
        uuid: '002',
        url: 'one',
        child_uuid: '',
        parent_uuid: '001'
      },
      {
        type: 'TITLE',
        title: 'Title2',
        uuid: '003',
        child_uuid: '004',
        parent_uuid: ''
      },
      {
        type: 'DOC',
        title: '文档2',
        uuid: '004',
        url: 'two',
        child_uuid: '',
        parent_uuid: '003'
      }
    ]
    const uuidMap = new Map()
    const pr = new ProgressBar(testTools.cwd, tocList.length)
    await pr.init()

    await downloadArticleList({
      articleUrlPrefix: 'https://www.yuque.com/yuque/base1',
      total: tocList.length,
      uuidMap,
      tocList: tocList as any,
      bookPath: testTools.cwd,
      bookId: 1111,
      host: 'https://www.yuque.com',
      imageServiceDomains: ['gxr404.com'],
      progressBar: pr,
      options: {
        ignoreImg: false
      } as any,
    })
    expect(pr.curr).toBe(tocList.length)
    let doc1Data = readFileSync(path.join(testTools.cwd, tocList[0].title, `${tocList[1].title}.md`)).toString()
    doc1Data = doc1Data.replace(/\.\/img.*?-(\d{6})\./g, (match, random) => {
      return match.replace(random, '123456')
    })
    expect(doc1Data).toMatchSnapshot()
    const imgPath = path.join(testTools.cwd, tocList[0].title, 'img', tocList[1].uuid)
    const imgList = readdirSync(imgPath)
    expect(readFileSync(path.join(imgPath, imgList[0])).length).toBe(99892)
    expect(readFileSync(path.join(imgPath, imgList[1])).length).toBe(81011)

    const doc2Data = readFileSync(path.join(testTools.cwd, tocList[2].title, `${tocList[3].title}.md`)).toString()
    expect(doc2Data).toMatchSnapshot()
  })
  it('the title is also a doc', async () => {
    const tocList = [
      {
        type: 'DOC',
        title: 'Title1_文档',
        uuid: '001',
        url: 'one',
        child_uuid: '002',
        parent_uuid: ''
      },
      {
        type: 'DOC',
        title: '文档1',
        uuid: '002',
        url: 'two',
        child_uuid: '',
        parent_uuid: '001'
      }
    ]
    const uuidMap = new Map()
    const pr = new ProgressBar(testTools.cwd, tocList.length)
    await pr.init()

    await downloadArticleList({
      articleUrlPrefix: 'https://www.yuque.com/yuque/base1',
      total: tocList.length,
      uuidMap,
      tocList: tocList as any,
      bookPath: testTools.cwd,
      bookId: 1111,
      host: 'https://www.yuque.com',
      imageServiceDomains: ['gxr404.com'],
      progressBar: pr,
      options: {
        ignoreImg: false
      } as any,
    })
    expect(pr.curr).toBe(tocList.length)

    let doc1Data = readFileSync(path.join(testTools.cwd, tocList[0].title, 'index.md')).toString()
    doc1Data = doc1Data.replace(/\.\/img.*?-(\d{6})\./g, (match, random) => {
      return match.replace(random, '123456')
    })
    expect(doc1Data).toMatchSnapshot()
    const imgPath = path.join(testTools.cwd, tocList[0].title, 'img', tocList[0].uuid)
    const imgList = readdirSync(imgPath)
    expect(readFileSync(path.join(imgPath, imgList[0])).length).toBe(99892)
    expect(readFileSync(path.join(imgPath, imgList[1])).length).toBe(81011)

    const doc2Data = readFileSync(path.join(testTools.cwd, tocList[0].title, `${tocList[1].title}.md`)).toString()
    expect(doc2Data).toMatchSnapshot()
  })

})
