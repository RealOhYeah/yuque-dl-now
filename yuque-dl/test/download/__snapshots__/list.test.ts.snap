// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`downloadArticle > should work 1`] = `
"# 文档1

# DOC1
![1.jpeg](./img/002/1-123456.jpeg)
## SubTitle
![2.jpeg](./img/002/2-123456.jpeg)

> 原文: <https://www.yuque.com/yuque/base1/one>"
`;

exports[`downloadArticle > should work 2`] = `
"# 文档2

# DOC2
## Title
text...
### Title2
text...


> 原文: <https://www.yuque.com/yuque/base1/two>"
`;

exports[`downloadArticle > the title is also a doc 1`] = `
"# Title1_文档

# DOC1
![1.jpeg](./img/001/1-123456.jpeg)
## SubTitle
![2.jpeg](./img/001/2-123456.jpeg)

> 原文: <https://www.yuque.com/yuque/base1/one>"
`;

exports[`downloadArticle > the title is also a doc 2`] = `
"# 文档1

# DOC2
## Title
text...
### Title2
text...


> 原文: <https://www.yuque.com/yuque/base1/two>"
`;
